pytorch_lightning/trainer/trainer.py

```diff
@@ -1431,58 +1437,84 @@ def _call_teardown_hook(self) -> None:
-            if callable(model_fx):
-                output = model_fx(*args, **kwargs)
-
-            # *Bad code alert*
-            # The `Accelerator` mostly calls the `TrainingTypePlugin` but some of those calls are deprecated.
-            # The following logic selectively chooses which hooks are called on each object.
-            # In the case of `setup` and `teardown`, the hooks on the `LightningModule` should not call the hooks of the
-            # same name in these objects as they are meant to be managed outside of the `LightningModule` lifecycle.
-            # All of this should be fixed by #8506
-
-            # call the accelerator hook
-            if hook_name in ("on_train_start",) and hasattr(self.accelerator, hook_name):
-                accelerator_hook = getattr(self.accelerator, hook_name)
-                accelerator_output = accelerator_hook(*args, **kwargs)
-                # Rely on the accelerator output if lightningModule hook returns nothing
-                # Required for cases such as DataParallel where we reduce the output for the user
-                # todo: move this data parallel logic into the data parallel plugin
-                output = accelerator_output if output is None else output
-
-            # call the ttp hook
-            if hook_name not in ("setup", "teardown", "on_train_start") and hasattr(
-                self.training_type_plugin, hook_name
-            ):
-                ttp_hook = getattr(self.training_type_plugin, hook_name)
-                ttp_output = ttp_hook(*args, **kwargs)
-                output = ttp_output if output is None else output
+            fn = getattr(pl_module, hook_name)
+            if callable(fn):
+                # TODO: when profiling separate hook name by LM hooks/Callback hooks
+                with self.profiler.profile(hook_name):
+                    output = fn(*args, **kwargs)
+
+            # restore current_fx when nested context
+            pl_module._current_fx_name = prev_fx_name
+
+        return output
+
+    def _call_callback_hooks(
+        self,
+        hook_name: str,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Optional[Any]:
+        pl_module = self.lightning_module
+        output = None
+
+        if pl_module:
+            prev_fx_name = pl_module._current_fx_name
+            pl_module._current_fx_name = hook_name
+
+        # TODO: remove if in v1.7
+        if hook_name in ("on_train_batch_start", "on_train_batch_end"):
+            fn = getattr(self, hook_name)
+            if callable(fn):
+                with self.profiler.profile(hook_name):
+                    output = fn(*args, **kwargs)
+        else:
+            for callback in self.callbacks:
+                if hook_name in ("on_init_start", "on_init_end"):
+                    # these `Callback` hooks are the only ones that do not take a lightning module.
+                    # we also don't profile bc profiler hasn't been set yet
```

Comment:
is this a minor optimization? the `PassThroughProfiler` should be set

Reply:
you are right for `on_init_end` the profiler has been set, but not for `on_init_start` since that happens here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/3d6262b7a91215e72019d720e742e6261e1636dc/pytorch_lightning/trainer/trainer.py#L502

and we configure profiler here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/3d6262b7a91215e72019d720e742e6261e1636dc/pytorch_lightning/trainer/trainer.py#L560